libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/AlternativeRouteProgressDataProvider.kt

```diff
@@ -0,0 +1,37 @@
+package com.mapbox.navigation.core.routealternatives
+
+import com.mapbox.navigation.core.RouteProgressData
+
+internal object AlternativeRouteProgressDataProvider {
+
+    fun getRouteProgressData(
+        primaryRouteProgressData: RouteProgressData,
+        alternativeMetadata: AlternativeRouteMetadata
+    ): RouteProgressData {
+        val legIndex: Int
+        val routeGeometryIndex: Int
+        val legGeometryIndex: Int?
+        val primaryFork = alternativeMetadata.forkIntersectionOfPrimaryRoute
+        val alternativeFork = alternativeMetadata.forkIntersectionOfAlternativeRoute
+        val isForkPointAheadOnPrimaryRoute =
+            primaryRouteProgressData.routeGeometryIndex < primaryFork.geometryIndexInRoute
+        if (isForkPointAheadOnPrimaryRoute) {
+            val legIndexDiff = primaryFork.legIndex - alternativeFork.legIndex
+            legIndex = primaryRouteProgressData.legIndex - legIndexDiff
+            val routeGeometryIndexDiff =
+                primaryFork.geometryIndexInRoute - alternativeFork.geometryIndexInRoute
+            routeGeometryIndex =
+                primaryRouteProgressData.routeGeometryIndex - routeGeometryIndexDiff
+            val legGeometryIndexDiff =
+                primaryFork.geometryIndexInLeg - alternativeFork.geometryIndexInLeg
+            legGeometryIndex = primaryRouteProgressData.legGeometryIndex?.let {
+                it - legGeometryIndexDiff
+            }
+        } else {
+            legIndex = alternativeFork.legIndex
+            routeGeometryIndex = alternativeFork.geometryIndexInRoute
+            legGeometryIndex = alternativeFork.geometryIndexInLeg
+        }
```

Comment:
We are not covering a case where the alternative route is _longer_ than the primary route.

This can happen if:
1. We start navigation.
2. Somewhere along the route we receive an alternative.
3. We click that alternative to switch to it, before the fork.

Now, we have an alternative (former primary route) that is longer than the primary route and refreshing that alternative only from the fork point fails to refresh everything between the fork and current location.

Reply:
You are right, I highlighted the wrong portion of the code, but the problem still stands.

What I mean is that (assuming we adopt https://github.com/mapbox/mapbox-navigation-android/pull/6848#pullrequestreview-1261681325), then the `legIndex` calculation remains wrong:
```
val legIndexDiff = primaryFork.legIndex - alternativeFork.legIndex
legIndex = primaryRouteProgressData.legIndex - legIndexDiff
```

In the scenario mentioned above (if a route has two legs, and we're already on the second leg with a fork point ahead of us), we'll have:
```
primaryFork.legIndex == 0
alternativeFork.legIndex == 1
primaryRouteProgressData.legIndex == 0
```
which will return a `legIndex == -1`.

Reply:
It will return `legIndex = 1`.

```
val legIndexDiff = primaryFork.legIndex - alternativeFork.legIndex = 0 - 1 = -1
legIndex = primaryRouteProgressData.legIndex - legIndexDiff = 0 - (-1) = 1
```

Reply:
Touché!

So the last missing piece that I see is the `legGeometryIndex`. My napkin calculation seems to point out that the code does not work if the fork point is not on the current leg.

The code:
```
val legGeometryIndexDiff =
    primaryFork.geometryIndexInLeg - alternativeFork.geometryIndexInLeg
legGeometryIndex = primaryRouteProgressData.legGeometryIndex?.let {
    it - legGeometryIndexDiff
}
```

If fork point is on the upcoming leg, the `primaryFork.geometryIndexInLeg  - alternativeFork.geometryIndexInLeg` will always return `0` and the `legGeometryIndex` will be equal to `primaryRouteProgressData.legGeometryIndex` which is incorrect. Please correct me if I'm wrong though!

Reply:
Yes, that's an issue. Fixed.

Reply:
We'll need a test for that as well. Looking at the setup we could take the existing `route_response_single_route_multileg` and `route_response_single_route_multileg_alternative` routes and swap the order to test both cases.

Reply:
Oh, you meant the instrumentation test. I'll add one.